### PR TITLE
Add `-AsPlainText` to `ConvertFrom-SecureString`

### LINKED ISSUE #11067: Feature Request: Convert SecureString to string
# Summary of the new feature/enhancement

It's becoming more common for scripts to have a secret as a secure string, but needs to use it as a plain text string (like making a REST call).  Anyone can search the web to find the APIs needed to convert the SecureString to plain text.  It seems we should just make this a part of PowerShell so people aren't blindly copying and pasting script samples from websites.

# Proposed technical implementation details (optional)

Proposal is to add `-AsPlainText` to `ConvertFrom-SecureString` cmdlet.

### PR DESCRIPTION
**Pull request by SteveL-MSFT**
# PR Summary

Fix https://github.com/PowerShell/PowerShell/issues/11067

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

**Review by vexx32 — APPROVED** [review]

File: src/Microsoft.PowerShell.Security/security/SecureStringCommands.cs [hunk 1]
```diff
@@ -139,6 +139,12 @@ public SecureString SecureString
             }
         }
 
+        /// <summary>
+        /// Gets or sets a switch to get the secure string as plain text.
+        /// </summary>
+        [Parameter(ParameterSetName = "AsPlainText")]
+        public SwitchParameter AsPlainText { get; set; }
```
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Security/security/SecureStringCommands.cs]: One minor thing: `ConvertTo-SecureString -AsPlainText` requires `-Force` to be specified:
```
PS> ConvertTo-SecureString -AsPlainText abc
ConvertTo-SecureString: The system cannot protect plain text input. To suppress this warning and convert the plain text to a SecureString, reissue the command specifying the Force parameter. For more information, type: get-help ConvertTo-SecureString.
```
It's more of a storngly-enforced warning. Shall we do the same to `ConvertFrom-SecureString -AsPlainText`?

Comment (SteveL-MSFT) [reply src/Microsoft.PowerShell.Security/security/SecureStringCommands.cs]: Let's discuss this in the issue

**Review by PaulHigin — APPROVED** [review]
LGTM

File: src/Microsoft.PowerShell.Security/security/SecureStringCommands.cs (hunk 1 above)
Comment (PaulHigin) [reply src/Microsoft.PowerShell.Security/security/SecureStringCommands.cs]: FWIW, I feel the `-Force` parameter is not needed, and we should consider removing it as a 'plaintext' requirement from `ConvertTo-SecureString`.

**Comment by TravisEz13** [comment]
should this be marked as rc.1-consider milestone?

**Comment by SteveL-MSFT** [comment]
Need to update PR based on @PowerShell/powershell-committee decision that `-Force` on `ConvertTo-SecureString` shouldn't be mandatory

**Comment by SteveL-MSFT** [comment]
Codacy appears to be wrong, VSCode complains if I remote the assignment since the finally block may use an unassigned variable.

**Comment by KirkMunro** [comment]
I think removing the hard requirement for `-Force` when using `ConvertTo-SecureString -AsPlainText` is a bad idea. I also think `ConvertFrom-SecureString -AsPlainText` should require `-Force` to use it.

The reason is really simple: someone looking for how to perform such a task in PowerShell is `-Force`d to consider the ramifications/risk and think twice about what they are trying to do (there may be a more secure way to do so).

Ease of use is less important than guiding scripters towards a better understanding of the security implications when working with plain-text passwords in scripts.

People are lazy. Make it too easy for them, and they'll take the easy way without thinking about why they shouldn't do that.

**Comment by rkeithhill** [comment]
What's the point on non-Windows platforms where "SecureString" is a bit of a lie anyway?  Requiring `-Force` on those platforms would be misleading IMO because those strings aren't really that secure.  See https://docs.microsoft.com/en-us/dotnet/api/system.security.securestring?view=netcore-3.0#HowSecure and the recommendation to not use this type anymore: https://github.com/dotnet/platform-compat/blob/master/docs/DE0001.md

That said, once we get the Secrets management module, I don't see the `*-SecureString` cmdlets being that useful anymore.

>  guiding scripters towards a better understanding of the security implications guiding scripters towards a better understanding of the security implications

Wouldn't PSScriptAnalyzer serve this purpose better?

**Comment by KirkMunro** [comment]
> What's the point on non-Windows platforms where "SecureString" is a bit of a lie anyway? Requiring `-Force` on those platforms would be misleading IMO because those strings aren't really that secure. See https://docs.microsoft.com/en-us/dotnet/api/system.security.securestring?view=netcore-3.0#HowSecure and the recommendation to not use this type anymore: https://github.com/dotnet/platform-compat/blob/master/docs/DE0001.md

In my mind it's not about `SecureString` and whether or not it is a bit of a lie since you can pull the text out of the string using APIs. It's about interactions between PowerShell and other things -- files on disk, credentials being sent in plain text over a connection, etc.

> That said, once we get the Secrets management module, I don't see the `*-SecureString` cmdlets being that useful anymore.

Agreed.
 
> > guiding scripters towards a better understanding of the security implications guiding scripters towards a better understanding of the security implications
> 
> Wouldn't PSScriptAnalyzer serve this purpose better?

PSScriptAnalyzer could and should raise warnings for various things related to plain text passwords, but that's only helpful if folks use PSSA. In my opinion, from a security/awareness perspective, pairing `-AsPlainText` with `-Force` is helpful.

**Comment by SteveL-MSFT** [comment]
My motivation for adding this capability is specifically for Secrets Management module.  Secrets Management will always return a SecureString type.  You would use this cmdlet to convert to plain text, if needed.  As for `-Force`, the discussion in @PowerShell/powershell-committee is that the noun `SecureString` is already an indicator that care should be taken when using these cmdlets.  The user already has to specify `-AsPlainText` to get plain text.  Having them also add a superfluous `-Force` is unnecessary and just adds to annoyance or worst invites users to add `Force = true` in PSDefaultParameters.

**Comment by rkeithhill** [comment]
> Secrets Management will always return a SecureString type

Even though the CoreFx team is clearly recommending not to use this type for new development?

**Comment by SteveL-MSFT** [comment]
@rkeithhill Until corefx provides a SensitiveString type to replace SecureString, we still need a type that doesn't implement `ToString()` and dump the contents accidentally to a transcript or the screen.

**Comment by KirkMunro** [comment]
> As for `-Force`, the discussion in @PowerShell/powershell-committee is that the noun `SecureString` is already an indicator that care should be taken when using these cmdlets. The user already has to specify `-AsPlainText` to get plain text. Having them also add a superfluous `-Force` is unnecessary and just adds to annoyance or worst invites users to add `Force = true` in PSDefaultParameters.

For the uncommon occurrence where you want to convert plain text into a `SecureString`, requiring `-Force` isn't superfluous. It's a safety precaution -- a reminder to think twice about what you're doing. And it only requires 6 extra characters.

When you say it's unnecessary and just adds to annoyance or invites users to add `$PSDefaultParameterValues['ConvertTo-SecureString:Force'] = $true`, do you have evidence that users are really annoyed by this or using `$PSDefaultParameterValues` to do that? There are zero open/closed issues in this GitHub repository complaining about `ConvertTo-SecureString` requiring `-Force` when it is used with `-AsPlainText`. I'm sure it might annoy a few people who don't want the reminder. I've spoken with others who appreciate the cmdlet drawing their attention to what they're doing.

IMHO the change in behavior of `-Force` is out of scope for this PR, unjustified, and undesirable.

**Comment by adityapatwardhan** [comment]
@TravisEz13 Is this ready to merge?

**Review by adityapatwardhan — APPROVED** [review]

**Pull request merged**